### PR TITLE
proto-loader-gen-types: Support nominal typing with type branding

### PR DESCRIPTION
**Pull request by LunaTK**
# TL;DR;

This PR introduces a boolean option `branded` for `proto-loader-gen-types`.

If the branded option is on, `interface` representing a Message has additional `__type` property, that has the `fullName` of originating `Message` as its value.

```proto
package test;

message FooRequest {
  string name = 1;
}
```

turns into

```ts
interface FooRequest {
  name: string;
  __type: '.test.FooRequest';
}

```
# Details

Since typescript only supports structural typing, if any two types has the exact same structure, they can be used interchangeably.

This makes codes like below totally valid.

```ts
interface FooRequest {
  name: string;
}

interface BarRequest {
  name: string;
}

function requestFoo(param: FooRequest) { ... }
function requestBar(param: BarRequest) { ... }

const fooArg: FooRequest = { name: "this is foo" };
requestBar(fooArg);  // no compile error
```

**Note** : since method signature generated by proto-loader-gen-types is too complicated, I will use much intuitive version(i.e., requestFoo, requestBar) here.

If this behavior is not what you want, typically you can fix this in typescript by using technique called type branding.

https://github.com/microsoft/TypeScript/blob/7b48a182c05ea4dea81bab73ecbbe9e013a79e99/src/compiler/types.ts#L693-L698

```ts
interface FooRequest {
  name: string;
  __type: 'FooRequest';
}

interface BarRequest {
  name: string;
  __type: 'BarRequest';
}

function requestFoo(param: FooRequest) { ... }
function requestBar(param: BarRequest) { ... }

const fooArg: FooRequest = { name: "this is foo" } as FooRequest;
requestBar(fooArg);  // error!
```

# Restricted vs Permissive

It is often cumbersome to use type assertion when you are providing the branded type.

For example, there is no additional effort required for using branded type returned from gRPC method, but creating a branded object requires type assertion.

```ts
interface FooRequest {
  name: string;
  __type: 'FooRequest';
}

interface FooResponse {
  age: number;
  __type: 'FooResponse';
}

function requestFoo(param: FooRequest): FooResponse { ... }

const res = requestFoo({} as FooRequest); // need type assertion to create branded object
const _res: FooResponse = res;  // no need for type assertion
```

I wonder if we need to include restricted interface only, or should we separate options for restricted and permissive interface respectively?

# Motivating materials
- [Nominal typing](https://basarat.gitbook.io/typescript/main-1/nominaltyping#using-interfaces)
- [Nominal typing in TypeScript source code](https://github.com/Microsoft/TypeScript/blob/7b48a182c05ea4dea81bab73ecbbe9e013a79e99/src/compiler/types.ts#L693-L698)
- [Type branding utility for TypeScript](https://github.com/piotrwitek/utility-types#brandt-u)

### CODE REVIEW

**Comment by LunaTK** [comment]
Thanks for the reply!
If you at least allow this as an optional feature, I believe that it would be welcomed by many users in future.

> The result is that the field is essentially a fiction in the type system that doesn't reflect the runtime objects.

Yes, this is indeed true because the purpose of type branding is to provide compile-time type compatibility checking, not for the runtime. 

And this can also be seen as an advantage, depending on the perspective, because it doesn't imply any runtime overhead.

> https://github.com/microsoft/TypeScript/blob/7b48a182c05ea4dea81bab73ecbbe9e013a79e99/src/compiler/types.ts#L697-L698
> Note: the brands are never actually given values.  At runtime they have zero cost.

Thus, keeping this special `__type` property in mind, library users should never use this property in runtime. So the type guard example you provided is a misleading use-case of branded types.

To recap, this is because the goal of branded type is to detect abnormal data flow in compile time, not for runtime type checking.


> My approach so far with this library has been to have the types reflect the runtime objects as closely as possible

I agree with your idea. Normally I also do not prefer type branding, but I recently encountered with some requirements which need type-level metadata to make use of type manipulation. In this case, I had no option but to give up 100% reflective typing in order to get type-level metadata. For such use cases where you have no other options, leaving type branding at least as an option might be the last hope.

And the last word if you allow, to relief the uncomfortableness of type assertion for users to make objects, I guess we can exclude permissive interfaces from type branding.

> The goal of generating two separate interfaces for each message type is to describe two separate things as narrowly as possible: the objects that users can pass to the library as input, and the objects that the library will output

Because objects that the library will output can be the automated starting point for the type branding.

**Comment by murgatroid99** [comment]
> Thus, keeping this special `__type` property in mind, library users should never use this property in runtime. So the type guard example you provided is a misleading use-case of branded types.

The problem I see is that users may incorrectly believe that the property is usable at runtime. I believe this can be resolved with a documentation comment in the generated code for each `__type` field that says something like:

> This field is a type brand and is not populated at runtime. Instances of this type should be created using type assertions.

---

> And the last word if you allow, to relief the uncomfortableness of type assertion for users to make objects, I guess we can exclude permissive interfaces from type branding.

I think a better approach here would be to have two separate options, one for permissive types and another for restrictive types. Maybe `--inputBranded` and `--outputBranded`.

File: packages/proto-loader/bin/proto-loader-gen-types.ts [hunk 1]
```diff
@@ -844,7 +845,6 @@ async function runScript() {
     .default('bytes', 'Buffer')
     .default('inputTemplate', `${templateStr}`)
     .default('outputTemplate', `${templateStr}__Output`)
-    .default('outputBranded', false)
```
Comment (LunaTK) [inline packages/proto-loader/bin/proto-loader-gen-types.ts]: If you add more than two defaults, typescript compiler will complain about excessive type instantiation.

image

I think this is because of the depth limitation of recursive type inference. Seems like we already reached the maximum number of inferable parameters.

To mitigate this I removed defaults for `inputBranded` and `outputBranded` in this commit, making them optional.

Although they would still be treated as falsy by default(undefined), help description will not tell that they will be treated as falsy by default. This might confuse users.

File: packages/proto-loader/README.md [hunk 2]
```diff
@@ -92,8 +92,12 @@ Options:
...
       --outputTemplate   Template for mapping output or "restricted" type names
                                                 [string] [default: "%s__Output"]
-      --branded          Emit property for branded type whose value is fullName
-                         of the Message               [boolean] [default: false]
+      --inputBranded     Output property for branded type for  "permissive"
+                         types with fullName of the Message as its value
+                                                                       [boolean]
+      --outputBranded    Output property for branded type for  "restricted"
+                         types with fullName of the Message as its value
+                                                                       [boolean]
```
Comment (LunaTK) [inline packages/proto-loader/README.md]: You can not tell if `inputBranded` and `outputBranded` would be treated as falsy by default.

**Review by LunaTK — COMMENTED** [review]
I updated the two points you mentioned.

While adding the boolean option with default, I got excessive type instantiation error.

Seems like we are already right under the maximum number of inferable options due to the limitation of tsc, if I am not doing something wrong :(

I demonstrated two workarounds. 
But indeed, requiring workaround doesn't make you pleasant 😭

File: packages/proto-loader/bin/proto-loader-gen-types.ts [hunk 3]
```diff
@@ -864,6 +864,14 @@ async function runScript() {
...
       }
     })
+    .option('inputBranded', {
+      boolean: true,
+      default: false,
+    })
+    .option('outputBranded', {
+      boolean: true,
+      default: false,
+    })
```
[the remark this answers is not in the document]
Comment (murgatroid99) [reply packages/proto-loader/bin/proto-loader-gen-types.ts:874]: Changing everything else to use `option` seems like a good way to handle this.

[the remark this answers is not in the document]
Comment (LunaTK) [reply packages/proto-loader/bin/proto-loader-gen-types.ts]: Fixed it.

Since the order was not exactly aligned with declaration order in the code and I don't know how exactly they were decided, 
I adjusted declaration order a bit to preserve the original printed order.

**Review by murgatroid99 — APPROVED** [review]
This looks good to me. Thank you for your contribution. There's just one more thing: please send another PR to document these options in [this file](https://github.com/grpc/proposal/blob/master/L70-node-proto-loader-type-generator.md similar to grpc/proposal#326 that documented another recent update

**Comment by LunaTK** [comment]
Great!
I made a PR for the proposal as well.

https://github.com/grpc/proposal/pull/342

**Comment by murgatroid99** [comment]
I published this in version 0.7.4
